pylint/checkers/classes.py

```diff
@@ -907,14 +907,23 @@ def _check_unused_private_members(self, node: astroid.ClassDef) -> None:
             found = False
             if not is_attr_private(function_def.name):
                 continue
-            for attribute in node.nodes_of_class(astroid.Attribute):
+            for attribute in (
+                a
+                for a in node.nodes_of_class(astroid.Attribute)
+                if a.attrname == function_def.name
+                and a.scope() != function_def  # We ignore recursive calls
+            ):
+                is_call = isinstance(attribute.expr, astroid.Call)
                 if (
-                    attribute.attrname == function_def.name
-                    and attribute.scope() != function_def  # We ignore recursive calls
+                    not is_call  # self.__attrname / node_name.__attrname
                     and attribute.expr.name in ("self", node.name)
+                    or is_call  # type(self).__attrname
+                    and attribute.expr.func.name == "type"
+                    and attribute.expr.args[0].name == "self"
```

Comment:
We can try to infer `type(self)`, I would guess that should work in almost all of those cases
```py
                if isinstance(attribute.expr, astroid.Call):
                    inferred = safe_infer(attribute.expr)
                    if (
                        isinstance(inferred, astroid.ClassDef)
                        and inferred.name == node.name
                    ):
                        break
```